enhancements/network/ovs-dynamic-cpu-pinning.md

```diff
@@ -0,0 +1,129 @@
+
+## Summary
+
+Allow OVS daemons to run on non reserved CPU if the network workload requires
+cycles and non-reserved CPU are idle.
+
+## Motivation
+
+OVS runs on reserved CPUs when an OCP node is configured with [cpuManagerPolicy](https://docs.openshift.com/container-platform/4.12/scalability_and_performance/using-cpu-manager.html) set to `static`, and the kubelet is configured with [`system-reserved` CPUs](https://docs.openshift.com/container-platform/4.12/nodes/nodes/nodes-nodes-resources-configuring.html).
+
+Typically the customers want to minimize the amount of reserved CPUs and leave 
+as many available CPUs as possible for business-related workloads.
+
+We have seen in some customer cases that it is possible to disrupt both the node
+and the cluster when the networking load rises, because OVS cannot keep up with 
+the flow and starts removing routes from its tables. Which just makes the situation 
+worse as the amount of upcalls (software based processing when a connection is not 
+present in the OVS tables) increases and increases the cpu load even further.
+
+### User Stories
+
+> As an Openshift cluster administrator
+> I want to make the most of the node's CPUs
+> So that the cluster can endure a high network workload.
+
+### Goals
+
+Every cluster node that is configured with static CPU manager policy and
+a low number of reserved CPU does not experience any excessive latency 
+increase while dealing with high network loads.
+
+### Non-Goals
+
+The logic used by Node Tuning Operator to decide if and when to enable the feature
+are out of this enhancement's scope.
+
+## Proposal
+
+CPU resources today are divided into two groups (slices in systemd terminology):
+- `system-reserved`: node components (e.g. kubelet) and system components (systemd and everything it monitors, including OVS)
+- `kube-reserved`: pod workloads
+
+The main idea to resolve the issue described above is to allow OVS 
+to spread out to other cpus outside the reserved cpu pool. However 
+a care must be taken to not interfere with cpus that are assigned 
+to latency sensitive workloads, that are part of [`kube-reserved` pool](https://docs.openshift.com/container-platform/4.12/nodes/nodes/nodes-nodes-resources-configuring.html#nodes-nodes-resources-configuring-about_nodes-nodes-resources-configuring).
+
+### Workflow Description
+
+The feature is triggered by Node Tuning Operator (NTO) when it detects the 
+performance conditions to turn it on. The logic NTO uses to activate this 
+feature is out of this enhancement's scope. 
+The following steps will be implemented to turn the feature on:
+- NTO's daemonset moves the ovs daemons to a systemd slice different than the default one. 
+  It lets them to spread over CPUs that are assigned to Kubernetes workload (`kube-reserved` slice).
+  - At the same time, NTO's daemonset creates a file in `/etc/openvswitch/enable_cpu_pinning` 
+    on every node where it wants to enable the pinning feature.
```

Comment:
My understanding is that when NTO detects the performance condition (high load), we want to stop pinning the OVS daemon to the reserved CPUs. This means we want to allow OVS daemon to run on reserved CPUs as well as on other CPUs, right? If this is true, naming this `enable_cpu_pinning` seems contrary to what is wanted because, i.e. we are lifting the restriction to run exclusively on reserved CPUs.

Reply:
Good point. I was thinking about it in terms of "enable dynamic CPU pinning", so it is worth finding a better name. Here are some suggestions:

| |ovnkube-node parameter|file|
|---|---|---|
|a|`--enable-dynamic-ovs-cpu-pinning`|`/etc/openvswitch/enable_dynamic_cpu_pinning`|
|b|`--manage-ovs-cpu-affinity`|`/etc/openvswitch/enable_cpu_affinity_management`|
|c|`--pin-ovs-cpu-affinity`|`/etc/openvswitch/enable_cpu_affinity_pinning`|

I left the `pinning` word in some proposals as the CPU affinity of OVS daemons is `pinned` to the affinity of ovnkube-node, but I'm open to other options.

@ricky-rav , @MarSik WDYT?

Reply:
I would suggest /etc/openvswitch/enable_dynamic_cpu_affinity and not having an ovnkube-node argument at all.

Reply:
I like `/etc/openvswitch/enable_dynamic_cpu_affinity`

Reply:
On the other hand this option enables explicit cpu pinning from the OVN-k perspective and this is needed to avoid interfering with guaranteed pods. Lifting the restriction is done via NTO/PAO.

Reply:
Yep, dynamic cpu affinity is fine.